README.md
[![Venafi](https://raw.githubusercontent.com/Venafi/.github/master/images/Venafi_logo.png)](https://www.venafi.com/)
[![MPL 2.0 License](https://img.shields.io/badge/License-MPL%202.0-blue.svg)](https://opensource.org/licenses/MPL-2.0)
![Community Supported](https://img.shields.io/badge/Support%20Level-Community-brightgreen)
![Compatible with TPP 17.3+ & VaaS](https://img.shields.io/badge/Compatibility-TPP%2017.3+%20%26%20VaaS-f9a90c) 
_**This open source project is community-supported.** To report a problem or share an idea, use
**[Issues](../../issues)**; and if you have a suggestion for fixing the issue, please include those details, too.
In addition, use **[Pull Requests](../../pulls)** to contribute actual bug fixes or proposed enhancements.
We welcome and appreciate all contributions. Got questions or want to discuss something with our team?
**[Join us on Slack](https://join.slack.com/t/venafi-integrations/shared_invite/zt-i8fwc379-kDJlmzU8OiIQOJFSwiA~dg)**!_

# Venafi-token Provider for HashiCorp Terraform

The `venafi-token` provider streamlines the process of rotating tokens from a Trust Protection Platform (TLSPDC) 
instance. It provides a resource that allows tokens to be managed and rotated when they expire or a custom refresh 
threshold (in days) is met.

## Requirements

### Protection of the terraform state file

Make sure that you are protecting your terraform state file as per the best practices by Hashicorp: [https://developer.hashicorp.com/terraform/language/state/sensitive-data](https://developer.hashicorp.com/terraform/language/state/sensitive-data).  
This is an important step to prevent data breaches or leaks of sensitive data like usernames, passwords, tokens, secrets, etc.

### Trust between Terraform and Trust Protection Platform

The Trust Protection Platform REST API (WebSDK) must be secured with a
certificate. Generally, the certificate is issued by a CA that is not publicly
trusted so establishing trust is a critical part of your setup.

Two methods can be used to establish trust. Both require the trust anchor
(root CA certificate) of the WebSDK certificate. If you have administrative
access, you can import the root certificate into the trust store for your
operating system. If you don't have administrative access, or prefer not to
make changes to your system configuration, save the root certificate to a file
in PEM format (e.g. /opt/venafi/bundle.pem) and include it using the
`trust_bundle` parameter of your Venafi provider.

## Setup

The `venafi-token` Provider for HashiCorp Terraform is an officially verified integration. As such, releases are 
published to the [Terraform Registry](https://registry.terraform.io/providers/Venafi/venafi/latest) where they are 
available for `terraform init` to automatically download whenever the provider is referenced by a configuration file.  
No setup steps are required to use an official release of this provider other than to download and install Terraform 
itself.

To use a pre-release or custom built version of this provider, manually install the plugin binary into
[required directory](https://www.terraform.io/docs/commands/init.html#plugin-installation) using the prescribed
[subdirectory structure](https://www.terraform.io/docs/configuration/provider-requirements.html#source-addresses)
that must align with how the provider is referenced in the `required_providers` block of the configuration file.


## Usage

1. Declare the `venafi-token` provider as required for your plan:
   ```terraform
   terraform {
     required_providers {
       venafi-token = {
         source  = "Venafi/venafi-token"
       }
     }
     required_version = ">= 0.13"
   }
   ```
2. Declare a credential resource in your terraform plan:
   ```terraform
   resource "venafi-token_credential" "example" {}
   ```

3. Import the credential resource.  
!> NOTE: It is very important that the resource is imported and not created. It holds sensitive data should not be displayed 
during the terraform plan.  
!> [Detailed documentation on how to build the import string can be found here.](https://github.com/Venafi/terraform-provider-venafi-token/blob/main/docs/resources/credential.md)
   ```sh
   terraform import venafi-token_credential.example 'url=<value>,trust_bundle=<value>,refresh_token=<value>'
   ```
4. Assign the `access token` from `venafi-token_credential.example` to the venafi provider block:
   ```terraform
   provider "venafi" {
     alias        = "tpp_token"
     url          = var.TPP_URL
     zone         = var.TPP_ZONE
     trust_bundle = file(var.TRUST_BUNDLE)
     access_token = venafi-token_credential.example.access_token
   }
   ```
5. Run your terraform plan:
   ```shell
   terraform apply
   ```
   
## License

Copyright &copy; Venafi, Inc. All rights reserved.

This solution is licensed under the Mozilla Public License, Version 2.0. See [LICENSE](./LICENSE) for the full license text.

Please direct questions/comments to opensource@venafi.com.


CHANGELOG.md
## 0.1.1 (October 14th, 2023)
- Fixed crash issue when no trust bundle was passed in the credential resource import string. ([#12](https://github.com/Venafi/terraform-provider-venafi-token/issues/12))

## 0.1.0 (Occtober 1, 2023)
- Initial release
- Added ability to refresh TLSPDC access tokens
- Added three authentication methods: refresh token, client certificate and username/password (deprecated)

scripts/changelog-links.sh
#!/bin/bash

# This script rewrites [GH-nnnn]-style references in the CHANGELOG.md file to
# be Markdown links to the given github issues.
#
# This is run during releases so that the issue references in all of the
# released items are presented as clickable links, but we can just use the
# easy [GH-nnnn] shorthand for quickly adding items to the "Unrelease" section
# while merging things between releases.

set -e

if [[ ! -f CHANGELOG.md ]]; then
  echo "ERROR: CHANGELOG.md not found in pwd."
  echo "Please run this from the root of the terraform provider repository"
  exit 1
fi

if [[ `uname` == "Darwin" ]]; then
  echo "Using BSD sed"
  SED="sed -i.bak -E -e"
else
  echo "Using GNU sed"
  SED="sed -i.bak -r -e"
fi

PROVIDER_URL="https:\/\/github.com\/terraform-providers\/terraform-provider-token\/issues"

$SED "s/GH-([0-9]+)/\[#\1\]\($PROVIDER_URL\/\1\)/g" -e 's/\[\[#(.+)([0-9])\)]$/(\[#\1\2))/g' CHANGELOG.md

rm CHANGELOG.md.bak


internal/provider/provider.go
package provider

import (
	"context"

	"github.com/hashicorp/terraform-plugin-framework/datasource"
	"github.com/hashicorp/terraform-plugin-framework/provider"
	"github.com/hashicorp/terraform-plugin-framework/provider/schema"
	"github.com/hashicorp/terraform-plugin-framework/resource"
)

var _ provider.Provider = &VenafiTokenProvider{}

func New() provider.Provider {
	return &VenafiTokenProvider{}
}

type VenafiTokenProvider struct{}

func (p *VenafiTokenProvider) Metadata(_ context.Context, _ provider.MetadataRequest, resp *provider.MetadataResponse) {
	resp.TypeName = "venafi-token"
}

func (p *VenafiTokenProvider) Schema(_ context.Context, _ provider.SchemaRequest, resp *provider.SchemaResponse) {
	resp.Schema = schema.Schema{
		MarkdownDescription: "This is for refreshing Venafi tokens for use with venafi-provider.",
	}
}

func (p *VenafiTokenProvider) Configure(_ context.Context, _ provider.ConfigureRequest, _ *provider.ConfigureResponse) {

}

func (p *VenafiTokenProvider) DataSources(_ context.Context) []func() datasource.DataSource {
	return []func() datasource.DataSource{}
}

func (p *VenafiTokenProvider) Resources(_ context.Context) []func() resource.Resource {
	return []func() resource.Resource{
		NewCredentialResource,
	}
}


internal/vcertclient/client.go
// Package vcertclient contains all functions that interface with vcert-sdk
package vcertclient

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"encoding/pem"
	"fmt"
	"net/http"
	"os"

	"github.com/Venafi/vcert/v5"
	"github.com/Venafi/vcert/v5/pkg/endpoint"
	"github.com/Venafi/vcert/v5/pkg/venafi/tpp"
	"github.com/hashicorp/terraform-plugin-log/tflog"
	"golang.org/x/crypto/pkcs12"

	"github.com/terraform-providers/terraform-provider-venafi-token/internal/model"
)

const (
	msgTokenRefreshStart   = "retrieving token pair using"
	msgTokenRefreshSuccess = "successfully retrieved new token pair"
	msgTokenRefreshFail    = "failed to retrieve new token pair with"
	msgVcertClientError    = "terraform vcert client error"
)

type Client struct {
	context  context.Context
	credData model.CredentialResourceData
}

type RefreshTokenResponse struct {
	AccessToken  string
	RefreshToken string
	Expires      int64
	ExpiresIn    int64
}

func New(ctx context.Context, data model.CredentialResourceData) *Client {
	return &Client{
		context:  ctx,
		credData: data,
	}
}

func (c *Client) VerifyTokenExpired() (expired bool, err error) {
	tflog.Info(c.context, "verifying access token validity")

	config, err := c.createVCertConfig()
	if err != nil {
		tflog.Error(c.context, err.Error())
		return false, err
	}

	vClient, err := vcert.NewClient(config, false)
	if err != nil {
		tflog.Error(c.context, err.Error())
		return false, err
	}

	auth := &endpoint.Authentication{
		AccessToken: c.credData.AccessToken.ValueString(),
	}

	//Due to limitations in TPP API, we cannot retrieve the access token expiration time from the verify function
	_, err = vClient.(*tpp.Connector).VerifyAccessToken(auth)
	if err != nil {
		msg := fmt.Sprintf("%s: %s", msgVcertClientError, err.Error())
		tflog.Info(c.context, msg)
		return true, nil
	}

	return false, nil
}

func (c *Client) RequestNewTokenPair() (*RefreshTokenResponse, error) {
	tflog.Info(c.context, "requesting new token pair")

	tokenMethod := !c.credData.RefreshToken.IsNull()
	p12Method := !c.credData.P12Certificate.IsNull() && !c.credData.P12Password.IsNull()
	userMethod := !c.credData.Username.IsNull() && !c.credData.Password.IsNull()

	if !tokenMethod && !p12Method && !userMethod {
		return nil, fmt.Errorf("%s: no authorization methods specified", msgVcertClientError)
	}

	if tokenMethod {
		tflog.Info(c.context, fmt.Sprintf("%s %s", msgTokenRefreshStart, "refresh token"))
		resp, err := c.refreshAccessToken()
		// return if no errors
		if err == nil {
			tflog.Info(c.context, msgTokenRefreshSuccess)
			return resp, nil
		}
		// if refresh token fails. Check if there is any other auth method.
		// if there is another auth method, log warning and continue
		msg := fmt.Sprintf("%s %s: %s", msgTokenRefreshFail, "refresh token", err.Error())
		if !p12Method && !userMethod {
			// no other auth method. Log and return error
			tflog.Error(c.context, msg)
			return nil, fmt.Errorf("%s: %w", msgVcertClientError, err)
		}
		// print warning and let other auth methods be used
		tflog.Warn(c.context, msg)
	}

	if p12Method {
		tflog.Info(c.context, fmt.Sprintf("%s %s", msgTokenRefreshStart, "client certificate"))
		resp, err := c.getAccessTokenByP12()
		// return if no errors
		if err == nil {
			tflog.Info(c.context, msgTokenRefreshSuccess)
			return resp, nil
		}
		// if client certificate fails. Check if there is user/password method, log warning and continue
		msg := fmt.Sprintf("%s %s: %s", msgTokenRefreshFail, "client certificate", err.Error())
		if !userMethod {
			// no other auth method. Log and return error
			tflog.Error(c.context, msg)
			return nil, fmt.Errorf("%s: %w", msgVcertClientError, err)
		}
		// log warning and let other auth methods be used
		tflog.Warn(c.context, msg)
	}

	if userMethod {
		tflog.Info(c.context, fmt.Sprintf("%s %s", msgTokenRefreshStart, "username-password"))
		resp, err := c.getAccessTokenByUsernamePassword()
		// return if no errors
		if err == nil {
			tflog.Info(c.context, msgTokenRefreshSuccess)
			return resp, nil
		}
		// no other auth method. Log and return error
		tflog.Error(c.context, fmt.Sprintf("%s %s: %s", msgTokenRefreshFail, "username-password", err.Error()))
		return nil, fmt.Errorf("%s: %w", msgVcertClientError, err)
	}

	return nil, fmt.Errorf("%s: could not complete refresh token operation: all authentication methods failed", msgVcertClientError)
}

func (c *Client) RevokeToken() error {
	tflog.Info(c.context, "revoking access token")

	config, err := c.createVCertConfig()
	if err != nil {
		tflog.Error(c.context, err.Error())
		return err
	}

	vClient, err := vcert.NewClient(config, false)
	if err != nil {
		tflog.Error(c.context, err.Error())
		return err
	}

	auth := &endpoint.Authentication{
		AccessToken: c.credData.AccessToken.ValueString(),
	}
	err = vClient.(*tpp.Connector).RevokeAccessToken(auth)
	if err != nil {
		tflog.Error(c.context, err.Error())
		return err
	}

	return nil
}

func (c *Client) refreshAccessToken() (*RefreshTokenResponse, error) {
	tflog.Info(c.context, "using refresh token authentication method")

	config, err := c.createVCertConfig()
	if err != nil {
		return nil, err
	}
	vClient, err := vcert.NewClient(config, false)
	if err != nil {
		return nil, err
	}

	auth := &endpoint.Authentication{
		RefreshToken: c.credData.RefreshToken.ValueString(),
		ClientId:     c.credData.ClientID.ValueString(),
	}
	resp, err := vClient.(*tpp.Connector).RefreshAccessToken(auth)
	if err != nil {
		return nil, err
	}

	refreshResp := RefreshTokenResponse{
		AccessToken:  resp.Access_token,
		RefreshToken: resp.Refresh_token,
		Expires:      int64(resp.Expires),
	}

	return &refreshResp, nil
}

func (c *Client) getAccessTokenByP12() (*RefreshTokenResponse, error) {
	tflog.Info(c.context, "using client certificate authentication method")

	err := c.configureTLSClient()
	if err != nil {
		return nil, err
	}

	return c.getAccessToken(true)
}

func (c *Client) getAccessTokenByUsernamePassword() (*RefreshTokenResponse, error) {
	tflog.Info(c.context, "using username-password authentication method")

	return c.getAccessToken(false)
}

func (c *Client) getAccessToken(useClientCertificate bool) (*RefreshTokenResponse, error) {
	config, err := c.createVCertConfig()
	if err != nil {
		return nil, err
	}

	vClient, err := vcert.NewClient(config, false)
	if err != nil {
		return nil, err
	}

	auth := &endpoint.Authentication{
		ClientId: c.credData.ClientID.ValueString(),
	}

	if useClientCertificate {
		auth.ClientPKCS12 = true
	} else {
		auth.User = c.credData.Username.ValueString()
		auth.Password = c.credData.Password.ValueString()
	}

	resp, err := vClient.(*tpp.Connector).GetRefreshToken(auth)
	if err != nil {
		return nil, err
	}

	refreshResp := RefreshTokenResponse{
		AccessToken:  resp.Access_token,
		RefreshToken: resp.Refresh_token,
		Expires:      int64(resp.Expires),
	}
	return &refreshResp, nil
}

func (c *Client) configureTLSClient() error {
	tflog.Info(c.context, "configuring TLS client")

	p12Location := c.credData.P12Certificate.ValueString()
	password := c.credData.P12Password.ValueString()

	data, err := os.ReadFile(p12Location)
	if err != nil {
		return fmt.Errorf("%s: unable to read PKCS#12 file at [%s]: %w", msgVcertClientError, p12Location, err)
	}

	// We have a PKCS12 file to use, set it up for cert authentication
	blocks, err := pkcs12.ToPEM(data, password)
	if err != nil {
		return fmt.Errorf("%s: failed converting PKCS#12 archive file to PEM blocks: %w", msgVcertClientError, err)
	}

	var pemData []byte
	for _, b := range blocks {
		pemData = append(pemData, pem.EncodeToMemory(b)...)
	}

	// Construct TLS certificate from PEM data
	cert, err := tls.X509KeyPair(pemData, pemData)
	if err != nil {
		return fmt.Errorf("%s: failed reading PEM data to build X.509 certificate: %w", msgVcertClientError, err)
	}

	caCertPool := x509.NewCertPool()
	caCertPool.AppendCertsFromPEM(pemData)

	// Setup TLS configuration
	tlsConfig := tls.Config{
		Renegotiation: tls.RenegotiateFreelyAsClient,
		Certificates:  []tls.Certificate{cert},
		RootCAs:       caCertPool,
	}

	// Create own Transport to allow HTTP1.1 connections
	transport := &http.Transport{
		// Only one request is made with a client
		DisableKeepAlives: true,
		// This is to allow for http1.1 connections
		ForceAttemptHTTP2: false,
		TLSClientConfig:   &tlsConfig,
	}

	//Setting Default HTTP Transport
	http.DefaultTransport = transport

	tflog.Info(c.context, "TLS client configured")
	return nil
}

func (c *Client) createVCertConfig() (*vcert.Config, error) {
	config := vcert.Config{
		ConnectorType: endpoint.ConnectorTypeTPP,
		BaseUrl:       c.credData.URL.ValueString(),
		LogVerbose:    true,
	}

	if !c.credData.TrustBundle.IsNull() {
		location := c.credData.TrustBundle.ValueString()
		data, err := os.ReadFile(location)
		if err != nil {
			return nil, fmt.Errorf("%s: unable to read trust bundle file at [%s]: %w", msgVcertClientError, location, err)
		}
		config.ConnectionTrust = string(data)
	}

	return &config, nil
}


internal/provider/credential_modifier.go
package provider

import (
	"context"
	"fmt"
	"time"

	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
	"github.com/hashicorp/terraform-plugin-framework/types"
	"github.com/hashicorp/terraform-plugin-log/tflog"

	"github.com/terraform-providers/terraform-provider-venafi-token/internal/model"
	"github.com/terraform-providers/terraform-provider-venafi-token/internal/vcertclient"
)

var _ planmodifier.String = &credentialUnknownModifier{}

type credentialUnknownModifier struct{}

func (m credentialUnknownModifier) Description(_ context.Context) string {
	return "allow token refresh before token expires."
}

func (m credentialUnknownModifier) MarkdownDescription(_ context.Context) string {
	return "allow token refresh before token expires."
}

func (m credentialUnknownModifier) PlanModifyInt64(ctx context.Context, req planmodifier.Int64Request, resp *planmodifier.Int64Response) {
	tflog.Info(ctx, fmt.Sprintf("running venafi_credential plan modifier for %s", req.Path.String()))
	var data model.CredentialResourceData
	diags := req.Plan.Get(ctx, &data)
	resp.Diagnostics.Append(diags...)
	if resp.Diagnostics.HasError() {
		return
	}

	// Mark resource for update if there is no access token
	if data.AccessToken.IsNull() {
		tflog.Info(ctx, "no access token, marking resource for update")
		resp.PlanValue = types.Int64Unknown()
		return
	}

	client := vcertclient.New(ctx, data)
	expired, expirationDate, err := client.VerifyTokenExpired()
	if err != nil {
		tflog.Error(ctx, fmt.Sprintf("client error: %s", err.Error()))
		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to verify token, got error: %s", err))
		return
	}

	tflog.Info(ctx, fmt.Sprintf("token state: expired=%t, expirationDate=%d", expired, expirationDate))

	// If token already expired, mark resource for update
	if expired {
		tflog.Info(ctx, "access token expired, marking resource for update")
		resp.PlanValue = types.Int64Unknown()
		return
	}

	// If token not expired, check expiration date is on refresh window. If so, mark resource for update
	refreshWindowSeconds := data.ExpirationDate.ValueInt64() * 24 * 60 * 60
	if *expirationDate < (time.Now().Unix() - refreshWindowSeconds) {
		tflog.Info(ctx, "access token expiration within refresh window, marking resource for update")
		resp.PlanValue = types.Int64Unknown()
	}
}

func (m credentialUnknownModifier) PlanModifyString(ctx context.Context, req planmodifier.StringRequest, resp *planmodifier.StringResponse) {
	tflog.Info(ctx, fmt.Sprintf("running venafi_credential plan modifier for %s", req.Path.String()))

	var data model.CredentialResourceData
	diags := req.Plan.Get(ctx, &data)
	resp.Diagnostics.Append(diags...)
	if resp.Diagnostics.HasError() {
		return
	}

	// Mark resource for update if there is no access token
	if data.AccessToken.IsNull() {
		tflog.Info(ctx, "no access token, marking resource for update")
		resp.PlanValue = types.StringUnknown()
		return
	}

	client := vcertclient.New(ctx, data)
	expired, expirationDate, err := client.VerifyTokenExpired()
	if err != nil {
		tflog.Error(ctx, fmt.Sprintf("client error: %s", err.Error()))
		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to verify token, got error: %s", err))
		return
	}
	tflog.Info(ctx, fmt.Sprintf("token state: expired=%t, expirationDate=%d", expired, expirationDate))

	// If token already expired, mark resource for update
	if expired {
		tflog.Info(ctx, "access token expired, marking resource for update")
		resp.PlanValue = types.StringUnknown()
		return
	}

	// If token not expired, check expiration date is on refresh window. If so, mark resource for update
	refreshWindowSeconds := data.ExpirationDate.ValueInt64() * 24 * 60 * 60
	if *expirationDate < (time.Now().Unix() - refreshWindowSeconds) {
		tflog.Info(ctx, "access token expiration within refresh window, marking resource for update")
		resp.PlanValue = types.StringUnknown()
	}
}


internal/model/credential_resource_data.go
// Package model contains all data models used across the provider
package model

import "github.com/hashicorp/terraform-plugin-framework/types"

// CredentialResourceData represents a credential resource
type CredentialResourceData struct {
	URL            types.String `tfsdk:"url"`
	Username       types.String `tfsdk:"username"`
	Password       types.String `tfsdk:"password"`
	P12Certificate types.String `tfsdk:"p12_cert_filename"`
	P12Password    types.String `tfsdk:"p12_cert_password"`
	AccessToken    types.String `tfsdk:"access_token"`
	RefreshToken   types.String `tfsdk:"refresh_token"`
	ClientID       types.String `tfsdk:"client_id"`
	ExpirationDate types.Int64  `tfsdk:"expiration"`
	TrustBundle    types.String `tfsdk:"trust_bundle"`
	RefreshWindow  types.Int64  `tfsdk:"refresh_window"`
}


internal/provider/credential_resource.go
package provider

import (
	"context"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/hashicorp/terraform-plugin-framework/resource"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
	"github.com/hashicorp/terraform-plugin-framework/types"
	"github.com/hashicorp/terraform-plugin-log/tflog"

	"github.com/terraform-providers/terraform-provider-venafi-token/internal/model"
	"github.com/terraform-providers/terraform-provider-venafi-token/internal/vcertclient"
)

const (
	// attributes of the resource
	fURL            = "url"
	fUsername       = "username"
	fPassword       = "password"
	fP12Cert        = "p12_cert_filename"
	fP12Password    = "p12_cert_password"
	fAccessToken    = "access_token"
	fRefreshToken   = "refresh_token"
	fClientID       = "client_id"
	fExpirationDate = "expiration"
	fTrustBundle    = "trust_bundle"
	fRefreshWindow  = "refresh_window"

	// messages
	msgCredentialResourceError = "credential resource error"
	msgImportFail              = "failed to import certificate resource"

	// default values
	defaultClientID      = "hashicorp-terraform-by-venafi"
	defaultRefreshWindow = 30 // in days

	resourceNameSuffix = "credential"
)

var (
	_ resource.Resource                = &CredentialResource{}
	_ resource.ResourceWithImportState = &CredentialResource{}
)

func NewCredentialResource() resource.Resource {
	return &CredentialResource{}
}

type CredentialResource struct{}

func (r *CredentialResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
	resp.TypeName = fmt.Sprintf("%s_%s", req.ProviderTypeName, resourceNameSuffix)
}

func (r *CredentialResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
	resp.Schema = schema.Schema{
		MarkdownDescription: "Venafi Credential Resource",

		Attributes: map[string]schema.Attribute{
			fURL: schema.StringAttribute{
				MarkdownDescription: "The Venafi TLSPDC URL. Example: https://tpp.venafi.example/vedsdk",
				Optional:            true,
				Computed:            true,
			},
			fUsername: schema.StringAttribute{
				MarkdownDescription: "Username to authenticate to TLSPDC and request a new token",
				Optional:            true,
				Computed:            true,
			},
			fPassword: schema.StringAttribute{
				MarkdownDescription: "Password to authenticate to TLSPDC and request a new token",
				Optional:            true,
				Computed:            true,
				Sensitive:           true,
			},
			fP12Cert: schema.StringAttribute{
				MarkdownDescription: "base64-encoded PKCS#12 keystore containing a vcert certificate, private key, and chain certificates to authenticate to TLSPDC",
				Optional:            true,
				Computed:            true,
			},
			fP12Password: schema.StringAttribute{
				MarkdownDescription: "Password for the PKCS#12 keystore declared in p12_cert",
				Optional:            true,
				Computed:            true,
				Sensitive:           true,
			},
			fAccessToken: schema.StringAttribute{
				MarkdownDescription: "Access token used for authorization to TLSPDC",
				Computed:            true,
				Sensitive:           true,
			},
			fRefreshToken: schema.StringAttribute{
				MarkdownDescription: "Token used to request a new token pair (access/refresh token) from a TLSPDC instance",
				Optional:            true,
				Computed:            true,
				Sensitive:           true,
			},
			fClientID: schema.StringAttribute{
				MarkdownDescription: "Application that will be using the token",
				Optional:            true,
				Computed:            true,
			},
			fExpirationDate: schema.Int64Attribute{
				MarkdownDescription: "Expiration date of the access token",
				Optional:            true,
				Computed:            true,
			},
			fTrustBundle: schema.StringAttribute{
				MarkdownDescription: "Use to specify a base64-encoded, PEM-formatted file that contains certificates to be trust anchors for all communications with the Venafi TLSPDC instance",
				Optional:            true,
				Computed:            true,
			},
			fRefreshWindow: schema.Int64Attribute{
				MarkdownDescription: "number of days before expiration where a token refresh should be done",
				Optional:            true,
				Computed:            true,
			},
		},
	}
}

func (r *CredentialResource) Create(_ context.Context, _ resource.CreateRequest, resp *resource.CreateResponse) {
	resp.Diagnostics.AddError(msgCredentialResourceError, "credential resource cannot be created, only imported.")
}

func (r *CredentialResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
	tflog.Info(ctx, "reading credential resource")
	var data model.CredentialResourceData
	diags := req.State.Get(ctx, &data)
	resp.Diagnostics.Append(diags...)
	if resp.Diagnostics.HasError() {
		return
	}

	// No access token, request a new pair right away
	if data.AccessToken.IsNull() {
		tflog.Info(ctx, "no access token, retrieving a new token pair")
		err := rotateToken(ctx, &data)
		if err != nil {
			reportClientError(ctx, err, resp)
			return
		}
		resp.State.Set(ctx, data)
		return
	}

	// Got access token, check expiration
	client := vcertclient.New(ctx, data)
	expired, err := client.VerifyTokenExpired()
	if err != nil {
		tflog.Error(ctx, fmt.Sprintf("client error: %s", err.Error()))
		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to verify token expiration, got error: %s", err))
		return
	}

	// If token already expired, request new pair
	if expired {
		tflog.Info(ctx, "access token expired, retrieving a new token pair")
		err = rotateToken(ctx, &data)
		if err != nil {
			reportClientError(ctx, err, resp)
			return
		}
		resp.State.Set(ctx, data)
		return
	}

	// Refresh window is in days, we need to convert it to seconds: n days * 24 hours * 60 minutes * 60 seconds
	refreshWindowSeconds := data.RefreshWindow.ValueInt64() * 24 * 60 * 60
	// If token not expired, check expiration date is on refresh window. If so, request new pair
	if data.ExpirationDate.ValueInt64()-refreshWindowSeconds < time.Now().Unix() {
		tflog.Info(ctx, "access token expiration within refresh window, retrieving a new token pair")
		err = rotateToken(ctx, &data)
		if err != nil {
			reportClientError(ctx, err, resp)
			return
		}

		resp.State.Set(ctx, data)
		return
	}

	// Token is valid, nothing to do here
	tflog.Info(ctx, "access token valid")
}

func (r *CredentialResource) Update(ctx context.Context, _ resource.UpdateRequest, _ *resource.UpdateResponse) {
	tflog.Info(ctx, "updating credential resource")
}

func (r *CredentialResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
	tflog.Info(ctx, "deleting credential resource")
	var state model.CredentialResourceData

	diags := req.State.Get(ctx, &state)
	resp.Diagnostics.Append(diags...)
	if resp.Diagnostics.HasError() {
		return
	}

	client := vcertclient.New(ctx, state)
	err := client.RevokeToken()
	if err != nil {
		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to delete credential resource: %s", err.Error()))
		return
	}

	resp.State.RemoveResource(ctx)
	tflog.Info(ctx, "successfully revoked access token")
}

func (r *CredentialResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
	tflog.Info(ctx, "importing credential resource")
	id := req.ID

	dataMap, err := getValuesMap(ctx, id)
	if err != nil {
		details := fmt.Sprintf("%s: %s", msgImportFail, err.Error())
		resp.Diagnostics.AddError(msgCredentialResourceError, details)
		return
	}
	tflog.Debug(ctx, fmt.Sprintf("field map: %v", dataMap))

	data := model.CredentialResourceData{}

	msg := "saving attribute to terraform state: [%s]=%s"
	if val, ok := dataMap[fURL]; ok {
		tflog.Info(ctx, fmt.Sprintf(msg, fURL, val))
		data.URL = types.StringValue(val)
	}
	if val, ok := dataMap[fUsername]; ok {
		tflog.Info(ctx, fmt.Sprintf(msg, fUsername, val))
		data.Username = types.StringValue(val)
	}
	if val, ok := dataMap[fPassword]; ok {
		tflog.Info(ctx, fmt.Sprintf(msg, fPassword, val))
		data.Password = types.StringValue(val)
	}
	if val, ok := dataMap[fP12Cert]; ok {
		tflog.Info(ctx, fmt.Sprintf(msg, fP12Cert, val))
		data.P12Certificate = types.StringValue(val)
	}
	if val, ok := dataMap[fP12Password]; ok {
		tflog.Info(ctx, fmt.Sprintf(msg, fP12Password, val))
		data.P12Password = types.StringValue(val)
	}
	if val, ok := dataMap[fAccessToken]; ok {
		tflog.Info(ctx, fmt.Sprintf(msg, fAccessToken, val))
		data.AccessToken = types.StringValue(val)
	}
	if val, ok := dataMap[fRefreshToken]; ok {
		tflog.Info(ctx, fmt.Sprintf(msg, fRefreshToken, val))
		data.RefreshToken = types.StringValue(val)
	}
	if val, ok := dataMap[fTrustBundle]; ok {
		tflog.Info(ctx, fmt.Sprintf(msg, fTrustBundle, val))
		data.TrustBundle = types.StringValue(val)
	}

	clientID := defaultClientID
	if val, ok := dataMap[fClientID]; ok {
		clientID = val
	}
	tflog.Info(ctx, fmt.Sprintf(msg, fClientID, clientID))
	data.ClientID = types.StringValue(clientID)

	refreshWindow := defaultRefreshWindow
	if val, ok := dataMap[fRefreshWindow]; ok {
		valInt, err := strconv.Atoi(val)
		if err != nil {
			details := fmt.Sprintf("%s: %s", msgImportFail, err.Error())
			resp.Diagnostics.AddError(msgCredentialResourceError, details)
			return
		}
		refreshWindow = valInt
	}
	tflog.Info(ctx, fmt.Sprintf(msg, fRefreshWindow, fmt.Sprintf("%d", refreshWindow)))
	data.RefreshWindow = types.Int64Value(int64(refreshWindow))

	tflog.Debug(ctx, fmt.Sprintf("data struct: %v", data))
	diags := resp.State.Set(ctx, &data)
	resp.Diagnostics.Append(diags...)
	if resp.Diagnostics.HasError() {
		return
	}
}

func getValuesMap(ctx context.Context, values string) (map[string]string, error) {

	dict := make(map[string]string)

	list := strings.Split(values, ",")
	for _, item := range list {
		key, value, found := strings.Cut(item, "=")
		if !found {
			msg := fmt.Sprintf("no separator found on value: %s", item)
			tflog.Info(ctx, msg)
			return nil, fmt.Errorf(msg)
		}
		tflog.Debug(ctx, fmt.Sprintf("credential field found: %s = %s", key, value))
		dict[key] = value
	}

	return dict, nil
}

func rotateToken(ctx context.Context, data *model.CredentialResourceData) error {
	client := vcertclient.New(ctx, *data)
	clientResp, err := client.RequestNewTokenPair()
	if err != nil {
		return err
	}

	data.AccessToken = types.StringValue(clientResp.AccessToken)
	data.ExpirationDate = types.Int64Value(clientResp.Expires)
	data.RefreshToken = types.StringValue(clientResp.RefreshToken)

	return nil
}

func reportClientError(ctx context.Context, err error, resp *resource.ReadResponse) {
	tflog.Error(ctx, fmt.Sprintf("client error: %s", err.Error()))
	resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to rotate token, got error: %s", err.Error()))
}
